File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.py
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        
        
        fs = len(arr)
        cntr = Counter(arr) 
        ans = tmp = 0 
        for n in sorted(cntr.values() , key= lambda x:-x):
            ans +=1 
            tmp += n
            if tmp >= fs/2:
                return ans 

        return ans 

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution {
    public int minSetSize(int[] arr) {
        

        int al = arr.length ; 
        HashMap<Integer,Integer> cntr = new HashMap<>(); 
        for(int a:arr) cntr.put(a, cntr.getOrDefault(a,0)+1 ) ; 

        int[] count= new int[al+1] ; 
        for(int freq: cntr.values()) count[freq]++ ; 

        int ans =0 , tmp  = 0 , freq = al ; 
        while( tmp < al/2){
            ans ++ ; 
            while(count[freq] ==0) freq -- ; 
            tmp += freq ; 
            count[freq] -- ; 
        }
        return ans ; 
    }
}